### ISSUE TITLE
client-side `identify` method

### ISSUE THREAD

**Issue by ryw** [TelescopeJS/meteor-events#2]
Kicking off some discussion...

---

Do we want to replicate the `identify` concept from other libraries (client-side), or skip that complexity?

Normally server side you're required to manually supply userId on each call, so I don't think it's a problem to require that on client.

**Comment by jshimko** [comment]
My thinking was that we could include all (or most) of [the methods that analytics.js has](https://segment.com/docs/libraries/analytics.js/) to try to keep a consistent API so that integrating the two libs is easier.  So `identify` for logged in and anonymous users (not entirely sure how that will work for the Events collection, but probably worth considering since anonymous user data will be in the mix too), `track`, and `page`.  The `trackLink`, and `trackForm` methods aren't really necessary because of how Meteor/Iron Router works.  They're only for standard webpages that would leave the page before the event could be tracked (in the case of clicking a link or submitting a form).  Not an issue in Meteor world.  The regular `track` method will do the trick for those.

So yeah, identify/track/page and use the same properties object format.  My thinking for identify is that we could figure out a way to track anonymous users via cookie and merge if/when logged in (like analytics.js does).  That may be more complicated than it's worth though.  However, if we don't do it, you end up with a ton of anonymous event tracking data that could have ended up getting tied to specific users if those users later created an account.  Without a merge, you'd lose the details of the user signup flow for every user.  That's kind of a bummer.  But again, I'm not sure how difficult that merge is without taking a stab at it.  I suppose we could investigate the analytics.js source for ideas.  

Thoughts?

**Comment by ryw** [comment]
Yeah here's a simple implementation of those three methods that I did a while back, using Meteor's Session to track the userID:
https://github.com/Differential/assistant2/blob/master/packages/blackbeard/lib/client/blackbeard.coffee

Didn't do anything for anonymous users, because I'm personally pretty focused on tracking things that go on after the sign up.

I do like the idea of tracking anonymous sessions via cookie, even though it's not 100% reliable (same person, two devices, deleting cookies, etc.)

**Comment by SachaG** [comment]
I guess we could track events using a unique ID for every new cookie, and then update events in the db once we manage to identify the user. The only tricky part would be knowing when to trigger the update (i.e. detecting the first even that's tracked with a userId).

**Comment by jshimko** [comment]
Maybe we could do the merge on `Accounts.onUserCreate()` if that user's anonymous id exists in the Events collection?  Or maybe not.  I think you can only call that once per app.

**Comment by SachaG** [comment]
The only problem is that this only runs on the server, which won't have access to the cookie. So maybe `Accounts.onLogin()` would be better?

**Comment by ryw** [comment]
We override `Accounts.createUser` here on both client/server and it seems to be working for us, even if it's been overriden elsewhere: https://github.com/usercycle/meteor-events/blob/master/init.coffee

**Comment by jshimko** [comment]
Ah yes.  `Accounts.onLogin` would be way better since anonymous browsing can happen repeatedly - even if a user has an account.

**Comment by ryw** [comment]
Good point Jeremy... and I've never used Accounts.onLogin() before - love discovering new hooks ;)
